### ISSUE TITLE
Influxdb output - fails to use DNS record with multiple ipv4

### ISSUE THREAD

**Issue by lblasc** [influxdata/telegraf#5080]
### Relevant telegraf.conf:
```
[[outputs.influxdb]]
  urls = ["https://dns_name_with_multiple_ipv4.com"]
...
```
### System info:
```
# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.1 LTS
Release:	18.04
Codename:	bionic

# telegraf --version
Telegraf 1.8.0 (git: HEAD 07d2f2d8)
```

### Steps to reproduce:

1. Setup multiple HTTP(s) load balancers with public ipv4s (in my case nginx http load balancer) and point them to single influxdb.
2. Use those public ipv4 addresses to setup single dns name (in my case aws route 53):
```
dns_name_with_multiple_ipv4.com.	29	IN	A	1.2.3.4
dns_name_with_multiple_ipv4.com.	29	IN	A	4.5.6.7
dns_name_with_multiple_ipv4.com.	29	IN	A	7.8.9.1
```
3. Setup simple TCP health checks, and remove failed HTTP(s) load balancers ipv4 from the DNS pool when TCP failure is detected (in my case route 53 health checks)
4. point telegraf influxdb output to https://dns_name_with_multiple_ipv4.com
5. simulate TCP failures to reproduce

### Expected behavior:

When ip address in use starts to have failed to TCP connections telegraf should resolve dns address again and fallback to next address in pool.

### Actual behavior:

Telegraf on startup caches first resolved ip address and it doesn't do any DNS fallback mechanism. I suspect it will only use pool of ip addresses provided in telegraf config file for fallback.

**Comment by glinton** [comment]
DNS caching is a host thing, can you tune your host to not hold onto the resolutions for as long? See [this discussion on go dns caching](https://github.com/golang/go/issues/24796#issuecomment-383716244)

**Comment by lblasc** [comment]
Hello @glinton,

Thanks for responding. This is not what I meant with this issue, your comment mentions issue which clarifies who is in charge of caching DNS records and it is clear host system should do this not the process (telegraf in this case).

I'm talking about telegraf fallback mechanism when influxdb output address is DNS record with multiple A/AAAA records, in my opinion telegraf should fallback on next available A/AAAA record when TCP error occurs and maybe do round-robin on available records, this is up to you to decide, but some fallback mechanism should be in place.

For example, if I create 3x DNS records with single a record:
```
dns_name_with_single_ip_1.com.	29	IN	A	1.2.3.4
dns_name_with_single_ip_2.com.	29	IN	A	4.5.6.7
dns_name_with_single_ip_3.com.	29	IN	A	7.8.9.1
```

And set influxdb output like this:
```
[[outputs.influxdb]]
  urls = [
    "https://dns_name_with_single_ip_1.com",
    "https://dns_name_with_single_ip_2.com",
    "https://dns_name_with_single_ip_3.com"
  ]
...
```

I suspect (didn't try) telegraf will take first address from the list and use it, in event of TCP error it will start to use second address and in the new event of TCP error it will take the third address, etc..
Same thing should happen if single DNS name have multiple A/AAAA records.

Multiple records on single DNS name is pretty common setup for HTTP endpoints, so I'm very surprised to stumble on this issue. We lost lot of metrics because of single HTTP load balancer failure. 

I hope this explanation is clear enough.

Thanks!

**Comment by glinton** [comment]
Telegraf 1.9.0 has better queuing that i believe may help minimize lost metrics if an endpoint is unavailable, you may want to look into that.

If you are using DNS to manage load balancing (as you mentioned, you have health checks in place that remove an a-record from the registry) it shouldn't be up to the client to further repeat the health check process for each a-record.

The initial lookup when connecting to the db would return a single ip (even if n a-records were resolved). The behavior would then be different from having a single domain with multiple a-records and listing multiple addresses in the `urls` key. According to the comment on the [influx output's write function](https://godoc.org/github.com/influxdata/telegraf/plugins/outputs/influxdb#InfluxDB.Write), "Write sends metrics to one of the configured servers, logging each unsuccessful. If all servers fail, return an error." Since there is only one effective ip returned, telegraf can't continue checking, but it may be possible to add in some re-connect functionality.

I can agree that a re-connect functionality would be nice. Depending on system configuration, it could re-resolve the domain and use a different ip when attempting the next write. I'm not certain the current method of handling failed writes due to endpoint[s] being offline, but it sounds like they are just getting dropped. I'll re-open this to remind me to look more into that.

**Comment by lblasc** [comment]
Hello @glinton,

Simple re-connect function will be enough in case all `urls` are tried an none of them work, just go back to the first one in the list, re-resolve and try to write metrics. Re-resolving will return random A/AAAA record.

We will upgrade to telegraf 1.9 and use multiple DNS names with single A record in `urls` until this is handled properly. It will partially solve this issue, at least we will have some fallback mechanism.

Thanks!

**Comment by mjf** [comment]
I would suggest *Telegraf* cache all addresses from such "multi-address" *DNS RRs* and do round-robin, random, *IP*-sorted or whatever other imaginable walk-throughs as a client while respecting the *RR*'s *TTL* (which could be overrideable). Options should exist to change the default behaviour, especialy:

1. perform *DNS* resolution on every request and rely completely on *DNS* server to do the round-robin cycling which is rather typical behaviour (meaning - do not cache *DNS* responses at all);
2. *ditto* but also do round-robin (or another) cycling on its own (for cases when *DNS* server doesn't do the cycling itself which is rather uncommon behaviour but might exist somewhere);
3. perform *DNS* resolution first on startup and behave like *DNS* cache respecting *DNS RR*'s *TTL* and resolve again if *TTL* expires;
4. *ditto* but also do round-robin cycling on every request on its own.

In fact few extra configuration options are needed to achieve all the combinations, *i.e.*:

```toml
dns_caching = true   # default: false - resolve once on startup
dns_resolve_every = "300s"   # optional - overrides TTL from cache or resolve every ... if caching is off
dns_cycling_method = "random"   # default: "first-only", can be: "round-robin", "random", "ip-sorted", etc.
```

It would be nice if those *DNS*-related setting would be set both globaly and for plugins where *Telegraf* acts as a client and is supposed to do *DNS* resolution.

**Comment by danielnelson** [comment]
I tested this out using a simple dnsmasq setup making sure to disable `files` in `/etc/nsswitch.conf` so the dns server is always queried.
```
;; ANSWER SECTION:
influx-rr.              15      IN      A       127.0.0.2
influx-rr.              15      IN      A       127.0.0.3
influx-rr.              15      IN      A       127.0.0.4
```

Running host shows the addresses are being rotated:
```
$ host influx-rr 
influx-rr has address 127.0.0.3
influx-rr has address 127.0.0.4
influx-rr has address 127.0.0.2
$ host influx-rr 
influx-rr has address 127.0.0.4
influx-rr has address 127.0.0.2
influx-rr has address 127.0.0.3
$ host influx-rr 
influx-rr has address 127.0.0.2
influx-rr has address 127.0.0.3
influx-rr has address 127.0.0.4
```

When I run Telegraf and shutdown the influxdb server the output is:
```
2018-12-18T23:44:10Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"", Flush Interval:10s
mem available=17779396608i,free=1737179136i 1545176660000000000
mem free=1752182784i,available=17794453504i 1545176670000000000
2018-12-18T23:44:40Z E! [outputs.influxdb] when writing to [http://influx-rr:8086]: Post http://influx-rr:8086/write?consistency=any&db=telegraf: dial tcp 127.0.0.2:8086: connect: connection refused
2018-12-18T23:44:40Z E! [agent] Error writing to output [influxdb]: could not write any address
mem free=1781985280i,available=17824350208i 1545176680000000000
2018-12-18T23:44:50Z E! [outputs.influxdb] when writing to [http://influx-rr:8086]: Post http://influx-rr:8086/write?consistency=any&db=telegraf: dial tcp 127.0.0.3:8086: connect: connection refused
2018-12-18T23:44:50Z E! [agent] Error writing to output [influxdb]: could not write any address
mem free=1781596160i,available=17823993856i 1545176690000000000
2018-12-18T23:45:00Z E! [outputs.influxdb] when writing to [http://influx-rr:8086]: Post http://influx-rr:8086/write?consistency=any&db=telegraf: dial tcp 127.0.0.4:8086: connect: connection refused
2018-12-18T23:45:00Z E! [agent] Error writing to output [influxdb]: could not write any address
mem available=17840893952i,free=1798467584i 1545176700000000000
2018-12-18T23:45:10Z E! [outputs.influxdb] when writing to [http://influx-rr:8086]: Post http://influx-rr:8086/write?consistency=any&db=telegraf: dial tcp 127.0.0.2:8086: connect: connection refused
2018-12-18T23:45:10Z E! [agent] Error writing to output [influxdb]: could not write any address
mem free=1765756928i,available=17807855616i 1545176710000000000
```

**Comment by danielnelson** [comment]
I think this is working as expected, please let me know if I'm overlooking something.
